### ISSUE TITLE
Unnecessary syscalls on bare repository

### ISSUE THREAD

**Issue by Nukesor** [libgit2/libgit2#4480]
### Reproduction steps
Diff any two commits on bare repository and observe behaviour with strace.

### Expected behavior
Don't query stat for files on a bare repository.

### Actual behavior
Stat checked for every file in diff, if it's a bare repository.

### Version of libgit2 (release number or SHA1)
Master 782402c271d5467f9890af8e808af938a93bc55f

### Operating system(s) tested
Arch Linux

I just tried diffing with the current `master` libgit2 build. While the other bug with unnecessary syscalls for `.gitattributes` and `attributes` was fixed, there are still some `stat` calls for files, which shouldn't be necessary on a bare repository.

I'm using pygit2 for my interaction with git repositories, but as far as I can see, this is actually a issue with libgit2.

Current  output:
```
[pid 15375] access("/etc/gitattributes", F_OK) = -1 ENOENT (No such file or directory)
[pid 15375] stat("/tmp/gitalizer/phoemur/slackbuilds-1/info/attributes", 0x7ffe5758e5f0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/README", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/python-webencodings.SlackBuild", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/python-webencodings.SlackBuild", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/python-webencodings.info", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/python-webencodings.info", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/slack-desc", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/slack-desc", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/README", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/README", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/python-webencodings.SlackBuild", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/python-webencodings.SlackBuild", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/python-webencodings.info", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/python-webencodings.info", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/slack-desc", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
[pid 15375] stat("python/python-webencodings/slack-desc", 0x7ffe5758e7e0) = -1 ENOENT (No such file or directory)
```

pygit2 issue:
https://github.com/libgit2/pygit2/issues/764

**Comment by pks-t** [comment]
Hi @Nukesor and thanks for the report. I just tried to reproduce the behaviour you're seeing, but from using our diff example code I'm not able to do so. All stat syscalls either refer to the gitconfig, the repo's references or its object store. I definitely cannot see any stats to the files inside of the bare repository.

Are you able to provide a minimal working example with libgit2 that reproduces your problem?

**Comment by ethomson** [comment]
Does your repository have `core.bare` set or just it merely lack a working directory?

**Comment by Nukesor** [comment]
A clone alle repositories with the --bare flag. The config looks like this:
```
[core]
	bare = true
	repositoryformatversion = 0
	filemode = true
[remote "origin"]
	url = https://github.com/Nukesor/pueue.git
	fetch = +refs/heads/*:refs/remotes/origin/*
[branch "master"]
	remote = origin
	merge = refs/heads/master
```

**Comment by ethomson** [comment]
Does the repository have a working directory despite having that set?

**Comment by Nukesor** [comment]
It's as bare as it get's
```
⌁ [nuke:/tmp … /Nukesor/pueue] master ± ls -al
total 12
drwxr-xr-x 6 nuke nuke 200 Jan 18 12:57 .
drwxr-xr-x 3 nuke nuke  60 Jan 18 12:57 ..
-rw-r--r-- 1 nuke nuke 235 Jan 18 12:57 config
-rw-r--r-- 1 nuke nuke  73 Jan 18 12:57 description
-rw-r--r-- 1 nuke nuke   0 Jan 18 12:57 FETCH_HEAD
-rw-r--r-- 1 nuke nuke  23 Jan 18 12:57 HEAD
drwxr-xr-x 2 nuke nuke  60 Jan 18 12:57 hooks
drwxr-xr-x 2 nuke nuke  60 Jan 18 12:57 info
drwxr-xr-x 4 nuke nuke  80 Jan 18 12:57 objects
drwxr-xr-x 5 nuke nuke 100 Jan 18 12:57 refs
```

**Comment by ethomson** [comment]
Thanks for the details.  The reason that I ask is that the presence of a workdir overrides `core.bare`.  Eg, if you `GIT_WORK_DIR=. git diff ...` (and the libgit2 equivalent) then I would expect this behavior.

Out of curiosity, are you opening your repository with `discover_repository` or `open_repository`?

**Comment by Nukesor** [comment]
I'm using the Python standart `Repository` constructor,  which calls [init_file_backend](https://github.com/libgit2/pygit2/blob/master/src/pygit2.c#L162).
`init_file_backend` itself uses [git_repository_open](https://github.com/libgit2/pygit2/blob/master/src/pygit2.c#L162)`open_repository`. Thereby I assume I'm using `open_repository`.

**Comment by Nukesor** [comment]
Ok. I just tested a little and found at least the lines of code causing this behaviour.
Minimal example:
```
#!/bin/env python3

from collections import deque
from pygit2 import clone_repository, Repository

queue = deque()
commit_hashes = set()

repo = clone_repository('https://github.com/nukesor/pueue', '/tmp/pueue', bare=True)
# This repo initialization is just for testing if the behaviour only occurs with `clone_repository`
# But this is not the responsible code.
repo = Repository('/tmp/pueue')

master_commit = repo.head.get_object()
queue.appendleft(master_commit)

while len(queue) > 0:
    commit = queue.pop()
    # Break if we already visited this tree node
    if commit.hex in commit_hashes:
        continue

    commit_hashes.add(commit.hex)
    if len(commit.parents) == 1:
        diff = commit.tree.diff_to_tree(commit.parents[0].tree)
        # This is responsible
        # If I comment these lines of code there are no additional syscalls.
        test = {
            'additions': diff.stats.insertions,
            'deletions': diff.stats.deletions,
        }


    [queue.appendleft(parent) for parent in commit.parents]
```

**Comment by Nukesor** [comment]
Ok.
`diff.stats` is a wrapper around the Diff object, which is intialized with `git_diff_get_stats`. As far as I can see there is no additional logic for stat objects in pygit2.
I'm really sorry I can't provide a minimal example with libgit2, I'm trying to dig deeper into the code, but my C++ is really rusty and I don't have enough time to relearn it at the moment :/.

Anyway thanks for your support and the great libraries :)!

**Comment by pks-t** [comment]
No problem at all, thanks for the pygit2 reproducer. I will try to find some time for this soon-ish.

**Comment by tiennou** [comment]
I think I've pinpointed the problem to `git_attr_path__init` (here)(https://github.com/libgit2/libgit2/blob/master/src/attr_file.c#L512). It seems `git_attr_get_many_with_session` completely overlooks we have a bare repository, but I'm not sure where the fix should be…

Stack trace :
```
* thread #1: tid = 0x1f04ce6, 0x000000010038d864 libgit2_clar`git_attr_path__init(info=0x00007fff5fbfe9c8, path="test/helper.py", base=0x0000000000000000, dir_flag=GIT_DIR_FLAG_UNKNOWN) + 452 at attr_file.c:512, queue = 'com.apple.main-thread', stop reason = step out
Return value: (bool) $1 = false

  * frame #0: 0x000000010038d864 libgit2_clar`git_attr_path__init(info=0x00007fff5fbfe9c8, path="test/helper.py", base=0x0000000000000000, dir_flag=GIT_DIR_FLAG_UNKNOWN) + 452 at attr_file.c:512
    frame #1: 0x000000010038a410 libgit2_clar`git_attr_get_many_with_session(values=0x00007fff5fbfeab0, repo=0x000061400000fe40, attr_session=0x000061300000cb58, flags=0, pathname="test/helper.py", num_attr=1, names=0x00007fff5fbfeaa8) + 288 at attr.c:123
    frame #2: 0x00000001003bb34c libgit2_clar`git_diff_driver_lookup(out=0x000061500000d638, repo=0x000061400000fe40, attrsession=0x000061300000cb58, path="test/helper.py") + 236 at diff_driver.c:368
    frame #3: 0x00000001003bc981 libgit2_clar`git_diff_file_content__init_from_diff(fc=0x000061500000d628, diff=0x000061300000cb40, delta=0x000060b0000116b0, use_old=true) + 241 at diff_file.c:105
    frame #4: 0x000000010041ab78 libgit2_clar`patch_generated_init(patch=0x000061500000d500, diff=0x000061300000cb40, delta_index=0) + 216 at patch_generate.c:122
    frame #5: 0x000000010041994a libgit2_clar`patch_generated_alloc_from_diff(out=0x00007fff5fbfebd8, diff=0x000061300000cb40, delta_index=0) + 90 at patch_generate.c:140
    frame #6: 0x0000000100419760 libgit2_clar`git_patch_generated_from_diff(patch_ptr=0x00007fff5fbfed00, diff=0x000061300000cb40, idx=0) + 320 at patch_generate.c:724
    frame #7: 0x0000000100418c1e libgit2_clar`git_patch_from_diff(out=0x00007fff5fbfed00, diff=0x000061300000cb40, idx=0) + 158 at patch.c:210
    frame #8: 0x00000001003c616d libgit2_clar`git_diff_get_stats(out=0x00007fff5fbfeec0, diff=0x000061300000cb40) + 381 at diff_stats.c:191
    frame #9: 0x00000001000df49f libgit2_clar`test_diff_bare__dotest + 2895 at bare.c:67
    frame #10: 0x00000001000599f5 libgit2_clar`clar_run_test(test=0x00000001007856e0, initialize=0x0000000100811bc8, cleanup=0x0000000100811bd8) + 357 at clar.c:222
    frame #11: 0x0000000100055572 libgit2_clar`clar_run_suite(suite=0x0000000100811bc0, filter=0x0000000000000000) + 994 at clar.c:286
    frame #12: 0x0000000100054e5d libgit2_clar`clar_parse_args(argc=3, argv=0x00007fff5fbff4a0) + 1917 at clar.c:362
    frame #13: 0x0000000100054677 libgit2_clar`clar_test_run + 39 at clar.c:428
    frame #14: 0x00000001001d57e6 libgit2_clar`main(argc=3, argv=0x00007fff5fbff4a0) + 150 at main.c:24
    frame #15: 0x00007fffb1b1d235 libdyld.dylib`start + 1
```

C test case (with less memory safety) : https://gist.github.com/tiennou/11626f00a7892d1e929ec6f459e37894

**Comment by pks-t** [comment]
Thanks @tiennou for digging in. I looked around a bit based on your description and think I've found the issue. I've created a possible fix in #4510 -- while it is not really correct, we cannot do any better without a lookup into the ODB for the commits that are about to be diffed. Anyway, the optimized behavior is exactly the same as the previous one, though without additional stat calls.

**Comment by Nukesor** [comment]
The issue seems to be fixed!

Thanks a lot @pks-t @tiennou 👍 :)!!
